Scripts/Spells: implement warrior spell Trauma.

HannibalRoG opened the pull request:
[//]: # (***************************)
[//]: # (** FILL IN THIS TEMPLATE **)
[//]: # (***************************)

**Changes proposed:**

-  Implement 2 more spells for the warrior class, Whirlwind and Trauma.

**Target branch(es):** 3.3.5/master

- [x] master

**Issues addressed:** Closes #  (insert issue tracker number)
Could not find any

**Tests performed:** (Does it build, tested in-game, etc.)
Compiled and tested.

HannibalRoG committed:
Implement Warrior spell whirlwind and trauma.

DoctorKraft commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -1090,6 +1092,50 @@ class spell_warr_sword_and_board : public SpellScriptLoader
+        {
+            PreventDefaultAction();
+            // Proc only with Slam & Whirlwind
+            if (!(eventInfo.GetDamageInfo()->GetSpellInfo()->SpellFamilyFlags[0] & 0x200000 || eventInfo.GetDamageInfo()->GetSpellInfo()->SpellFamilyFlags[1] & 0x4))
```
`spell_proc_event` are better for condition of proc

HannibalRoG replied:
You responded so fast that I was scared, I check your comments and fix them, OMG so fast....

DoctorKraft replied:
Sorry x) I have already debug this spell but i help you for understand the spell

HannibalRoG replied:
Your help is apreciated.

wonder2011 replied:
@DoctorKraft please can you provide and example to learn how access spell_proc_event in the form you are saying ?

DoctorKraft replied:
https://trinitycore.atlassian.net/wiki/display/tc/spell_proc_event
in this example you need only
entry = spellid
SpellFamilyName = 4 (Warrior)
After for SpellFamilyMaskX
you need checl SpellFamilyMaskX in SpellWork for slam and whirlwind after you add up and convert in decimal

DoctorKraft commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -1090,6 +1092,50 @@ class spell_warr_sword_and_board : public SpellScriptLoader
+
+            //Get 25% of damage from the spell casted (Slam & Whirlwind)
+            int32 damage = eventInfo.GetDamageInfo()->GetDamage() * 25.0 / 100.0;
+            GetCaster()->CastCustomSpell(eventInfo.GetActionTarget(), SPELL_WARRIOR_TRAUMA_EFFECT, &damage, NULL, NULL, true);
```
What about new amount if spell proc again ?
- spell say "Multiple uses accumulate increased damage."

wonder2011 replied:
How you interpet "Multiple uses accumulate increased damage." ??. Heach time spell proc you increment by 25% ?. For example 25%, next hit 50%, next hit 75%, etc ?

DoctorKraft replied:
If target as has aura trauma (dot) and you proc again on this target. You need give a new amount of damage

Shauren replied:
there is a specialized method to do exactly this - `Unit::GetRemainingPeriodicAmount`

Keader replied:
Use nullptr instead NULL

wonder2011 replied:
@HannibalRoG is working on this now.

DoctorKraft commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -1232,6 +1278,44 @@ class spell_warr_vigilance_trigger : public SpellScriptLoader
 
+// 1680 - Whirlwind 
+// Updated 7.1.5
+class spell_warr_whirlwind : public SpellScriptLoader
```
Duplicate, Trauma are a proc so need only check condition for proc (whirlwind & slam) after your script `spell_warr_trauma` makes the rest

offl commented on sql/updates/world/master/RoG_22_03_2017_00_world.sql:
```diff
@@ -0,0 +1,4 @@
+DELETE FROM `spell_script_names` WHERE `ScriptName` IN ('spell_warr_whirlwind','spell_warr_trauma');
+INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
+('1680', 'spell_warr_whirlwind'),
+('215538', 'spell_warr_trauma'),
```
Missing semicolon.

Keader commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -1232,6 +1278,44 @@ class spell_warr_vigilance_trigger : public SpellScriptLoader
+        {
+            if (GetCaster()->HasAura(SPELL_WARRIOR_TRAUMA))
+                //GetCaster()->CastSpell(GetHitUnit(), SPELL_WARRIOR_TRAUMA_EFFECT, true);
+                GetCaster()->CastSpell(GetHitUnit(), SPELL_WARRIOR_TRAUMA, true);
```
typo here?
You validate SPELL_WARRIOR_TRAUMA_EFFECT, and cast SPELL_WARRIOR_TRAUMA again? No make sense to me

Commit:
Update and rename RoG_22_03_2017_00_world.sql to 2017_99_99_99_world.sql

- replaced comma with semicolon at end
- removed unwanted apostrophes around integers

HannibalRoG committed:
Only Trauma need script in core, thanks gays... appreciated
Fix .sql properly

HannibalRoG committed:
Merge branch 'master' into Warrior_Whirlwind_and_Truma

HannibalRoG committed:
Merge branch 'Warrior_Whirlwind_and_Truma' of https://github.com/Edreisantafe/TrinityCore into Warrior_Whirlwind_and_Truma

# Conflicts:
#	sql/updates/world/master/2017_99_99_99_world.sql

HannibalRoG committed:
Fix space in sql.

HannibalRoG commented:
Thanks again for all the comments that help us improve on the preparation of the script, thank you.
We implemented the method that Shauren recommended and reviewed all the annotations made to the code.

DoctorKraft commented:
@HannibalRoG nice work, it's better 👍 Just check your target and caster because it's an aurascript, you can crash worldserver without ^^

HannibalRoG committed:
Fixes.

Commit:
SQL codestyle

- remove apostrophes around integer numbers
- standardise comment section for compatibility

Seyden commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -1090,6 +1091,53 @@ class spell_warr_sword_and_board : public SpellScriptLoader
+                    //Get the Remaining Damage from the aura (if exist)
+                    int32 remainingDamage = target->GetRemainingPeriodicAmount(target->GetGUID(), SPELL_WARRIOR_TRAUMA_EFFECT, SPELL_AURA_PERIODIC_DAMAGE);
+                    //Get 25% of damage from the spell casted (Slam & Whirlwind) plus Remaining Damage from Aura
+                    int32 damage = (eventInfo.GetDamageInfo()->GetDamage() * 25.0 / 100.0) + remainingDamage;
```
Don't hardcode the 25.0.. 

replace 25.0 with aurEff->GetAmount()

and for the 100.0, replace it with 100.0f

Seyden commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -1090,6 +1091,53 @@ class spell_warr_sword_and_board : public SpellScriptLoader
+        void HandleProc(AuraEffect const* /*aurEff*/, ProcEventInfo& eventInfo)
+        {
+            PreventDefaultAction();
+            if (Unit* caster = GetCaster())
```
Caster can't be null here..

the caster has this aura.. so it wouldn't proc if there is no caster

Seyden commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -1090,6 +1091,53 @@ class spell_warr_sword_and_board : public SpellScriptLoader
+
+        void HandleProc(AuraEffect const* /*aurEff*/, ProcEventInfo& eventInfo)
+        {
+            PreventDefaultAction();
```
For what do you need PreventDefaultAction? the aura doesn't trigger anything.

HannibalRoG committed:
Some fixes from comments

HannibalRoG committed:
Merge branch 'Warrior_Whirlwind_and_Truma' of https://github.com/Edreisantafe/TrinityCore into Warrior_Whirlwind_and_Truma

wonder2011 commented:
What about this pr ?

HannibalRoG committed:
Merge branch 'master' into Warrior_Whirlwind_and_Truma

# Conflicts:
#	src/server/scripts/Spells/spell_warrior.cpp

HannibalRoG commented:
4 integrations and 2 conflicts solved, when will this enter the project?

Seyden commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -1093,6 +1094,49 @@ class spell_warr_sword_and_board : public SpellScriptLoader
+                //Get the Remaining Damage from the aura (if exist)
+                int32 remainingDamage = target->GetRemainingPeriodicAmount(target->GetGUID(), SPELL_WARRIOR_TRAUMA_EFFECT, SPELL_AURA_PERIODIC_DAMAGE);
+                //Get 25% of damage from the spell casted (Slam & Whirlwind) plus Remaining Damage from Aura
+                int32 damage = (eventInfo.GetDamageInfo()->GetDamage() * aurEff->GetAmount() / 100.0f) + remainingDamage;
```
Use CalculatePct(eventInfo.GetDamageInfo()->GetDamage(), aurEff->GetAmount()) instead of / 100

HannibalRoG committed:
AppVeyor error.

tkrokli commented:
You may want to merge master branch into this branch again, in particular to include the build fix in `master` and maybe also trigger travis-ci to pick up this PR again.

HannibalRoG committed:
Fix in function CalculatePct.

HannibalRoG commented:
Ok, thanks @tkrolki i already merge. I hope that now is not delayed so much the inclusion, this PR has been more than a month awaiting review. Do not rush, but we must go forward.

Shauren commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -1093,6 +1094,48 @@ class spell_warr_sword_and_board : public SpellScriptLoader
+                //Get the Remaining Damage from the aura (if exist)
+                int32 remainingDamage = target->GetRemainingPeriodicAmount(target->GetGUID(), SPELL_WARRIOR_TRAUMA_EFFECT, SPELL_AURA_PERIODIC_DAMAGE);
+                //Get 25% of damage from the spell casted (Slam & Whirlwind) plus Remaining Damage from Aura
+                int32 damage = CalculatePct(eventInfo.GetDamageInfo()->GetDamage(), aurEff->GetAmount()) + remainingDamage;
```
The calculated amount is wrong here, value you pass to `CastCustomSpell` is PER TICK amount so you are making it deal 25% damage of Slam on each tick (`GetRemainingPeriodicAmount` returns the value already divided so just adding it to pre-tick value is correct)

HannibalRoG replied:
@shauren do you mean something like

int32 damage = int32(CalculatePct(eventInfo.GetDamageInfo()->GetDamage(), aurEff->GetAmount()) / aurEff->GetTotalTicks()) + remainingDamage;

Shauren replied:
There is SpellInfo::GetMaxTicks, use that

Shauren commented:
Also, where is the Whirlwind fix? (PR title)

HannibalRoG committed:
Fix extra damage from Trauma.

HannibalRoG reviewed (commented):
@shauren Waiting for you.. I made all change requested, need revision.

gsi-israel commented:
This pr look fine. I tested in-game and is working good !. Merge please 👍 ?

gsi-israel commented:
@HannibalRoG please send me a chat invitation to fidelhernandezsalazarARROBAgmail.com

Shauren commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -1093,6 +1094,48 @@ class spell_warr_sword_and_board : public SpellScriptLoader
+                //Get the Remaining Damage from the aura (if exist)
+                int32 remainingDamage = target->GetRemainingPeriodicAmount(target->GetGUID(), SPELL_WARRIOR_TRAUMA_EFFECT, SPELL_AURA_PERIODIC_DAMAGE);
+                //Get 25% of damage from the spell casted (Slam & Whirlwind) plus Remaining Damage from Aura
+                int32 damage = int32(CalculatePct(eventInfo.GetDamageInfo()->GetDamage(), aurEff->GetAmount()) / aurEff->GetTotalTicks()) + remainingDamage;
```
My mistake here, I said "yes" to wrong code you suggested.
This still doesn't work because `aurEff` in `OnEffectProc` hook is the effect of the proc aura itself (215538) and this spell is not periodic (no ticks) so `aurEff->GetTotalTicks()` is 1 here. The correct function to use would be `sSpellMgr->AssertSpellInfo(SPELL_WARRIOR_TRAUMA_EFFECT)->GetMaxTicks()`

HannibalRoG committed:
Fix Trauma ticks

HannibalRoG committed:
Merge branch 'master' into Warrior_Whirlwind_and_Truma

HannibalRoG commented:
Ok @Shauren i think this is what you want, review please.

Shauren commented:
merged

wonder2011 commented:
Why TC Developers don't fix some spell in free time lol?
